### PR TITLE
Beomsu/eagle3

### PR DESCRIPTION
**Pull request by beomsu-mobilint**
### 개요

EAGLE-3는 **Base model**(대형 LLM)과 **Draft model**(경량 예측 모델)을 조합한 speculative decoding 기법이다. Draft model이 트리 형태로 여러 후보 토큰을 빠르게 생성하고, Base model이 한 번의 forward로 전체 트리를 검증해 다수의 토큰을 한꺼번에 수용(accept)한다.

#### Step 1. Prefill + 첫 Draft Tree 생성
1. Prompt의 미처리 구간(`input_ids[:, base_seq_length:]`)을 **Base MXQ**에 chunked prefill
2. Prefill logits에서 다음 토큰 1개 선택 (greedy 또는 sampling)
3. Base의 hidden states를 **FC Projector MXQ**로 차원 변환 후 **Draft MXQ**에 전달
4. Draft model이 `topk_generate`로 **tree 형태 후보 토큰** 생성

#### Step 2. Base Model Tree Verification
1. Draft tree의 후보 토큰을 **Base MXQ**에 tree attention으로 한 번에 평가
2. 이전 step에서 accept된 토큰이 있으면 position_ids를 보정해 연속 시퀀스로 결합
3. 평가 결과로 tree의 모든 노드에 대한 logits + hidden states 획득

#### Step 3. Posterior 기반 Accept 길이 결정
- **Greedy 경로**: 각 leaf-to-root path에서 `argmax(logits) == draft_token`인 연속 일치 길이를 계산, 최장 경로 선택
- **Sampling 경로**: top-k 분포에서 확률적으로 accept 여부 판정, reject 시 해당 확률 재분배 후 다음 후보 시도

#### Step 4. 상태 갱신 + 다음 Draft Tree 생성
1. 선택된 분기의 accept 토큰을 `input_ids`에 append
2. `cache.accept_tokens`에 저장 (다음 tree_decoding에서 position 보정용)
3. Accept된 hidden states + 샘플링된 다음 토큰으로 새 draft tree 생성
4. EOS 또는 `max_new_tokens` 도달 시 종료

### 3개 MXQ 모델의 Input/Output Shape

| MXQ | Inputs | Outputs |
|-----|--------|---------|
| **Base** | ① `inputs_embeds` `[1,1,seq,hidden]`<br>② `attention_mask` `[1,1,seq,cache+seq]`<br>③ `position_embs` `[1,1,seq,pe_size]` | ① `hidden1` + `hidden2` + `hidden3` → concat → `[1,seq,hidden_concat]`<br>② `logits` `[1,1,seq,vocab]` |
| **Draft** | ① `hidden_states` `[1,1,seq,hidden_draft]`<br>② `inputs_embeds` `[1,1,seq,embed_dim]`<br>③ `attention_mask` `[1,1,seq,cache+seq]`<br>④ `position_embs` `[1,1,seq,pe_size]` | ① `layer_outputs` (hidden features)<br>② `logits` |
| **FC Projector** | ① `hidden_states` `[1,1,seq,hidden_base]` | ① projected `[1,1,seq,hidden_draft]` |

> Base의 hidden output이 3개로 분할되는 이유: MXQ 런타임의 출력 텐서 제약에 맞춰 hidden을 3-split으로 내보내고 host에서 concat한다.  
> Draft의 input `hidden_states`가 Base hidden과 차원이 다를 경우 FC Projector를 경유한다.

### Draft Tree 텐서 구조 예시 (`top_k=3, depth=2, total_tokens=5`)

Base prefill 결과 샘플된 토큰 `T0`를 root로, Draft model이 확장 후 cumulative log-prob 상위 5개를 선택한 결과:

```text
T0 (root, idx 0)
├── A (idx 1, depth 1)
│   └── D (idx 4, depth 2)  ← leaf
├── B (idx 2, depth 1)
│   └── E (idx 5, depth 2)  ← leaf
└── C (idx 3, depth 1)       ← leaf (depth 2 후보가 pruning됨)
```

> Depth 1에서 3개, Depth 2에서 `3×3 = 9`개 후보가 생성되지만, 전체 12개 중 cumulative score 상위 `total_tokens=5`개만 유지한다. 이 예시에서는 C의 하위 후보는 score가 낮아 pruning되었다.

```python
draft_tokens      = [[T0, A, B, C, D, E]]     # shape [1, total_tokens+1] = [1, 6]
```

**`tree_position_ids`** — 각 노드의 depth:
```python
tree_position_ids  = [0, 1, 1, 1, 2, 2]        # shape [6]
```

**`tree_mask`** — 각 노드가 자신 + 조상만 attend (float, `[1, 1, 6, 6]`):

```text
          T0   A    B    C    D    E
T0      [  1    0    0    0    0    0 ]
A       [  1    1    0    0    0    0 ]
B       [  1    0    1    0    0    0 ]
C       [  1    0    0    1    0    0 ]
D       [  1    1    0    0    1    0 ]
E       [  1    0    1    0    0    1 ]
```

**`retrieve_indices`** — 각 leaf의 root→leaf 경로 (`[leaf_count, max_depth+1] = [3, 3]`):
```python
retrieve_indices = [[0, 3, -1],
                    [0, 1,  4],
                    [0, 2,  5]]
```

### RoPE Position Embedding: 일반 LLM MXQ와 EAGLE-3의 차이

**일반 LLM MXQ**에서는 RoPE가 **MXQ 모델 내부에 사전 계산**되어 있다. 토큰이 순차적으로 들어오므로 cache position만 전달하면 런타임이 내부적으로 올바른 position의 cos/sin을 적용한다.

**EAGLE-3**에서는 이 방식을 사용할 수 없다. Tree decoding 시 같은 depth의 여러 노드가 **동일한 position을 공유**하고, accept 토큰 prepend로 **비순차적 position 배열**이 생기기 때문이다. 따라서 EAGLE-3에서는:

1. **Host(CPU)에서 RoPE position table을 사전 계산**해두고
2. 매 forward 호출 시 **position_ids에 맞는 RoPE 텐서를 직접 인덱싱**하여
3. MXQ `infer()`의 **입력 텐서로 명시적으로 전달**한다

#### Position Table 사전 계산 및 전달

```python
inv_freq = 1.0 / (base ** (arange(0, dim, 2) / dim))
freqs = einsum("i,j->ij", positions, inv_freq)
emb = cat([freqs, freqs], dim=-1)
cos, sin = emb.cos(), emb.sin()
# interleaved layout으로 재배치 후 position_table 구성
indices = position_ids.view(-1).numpy()
rope_input = position_table[indices][None, None, :, :]  # [1, 1, seq, pe_size]
```

### Cache (`MobilintEagle3Cache`) 매 Step 업데이트 상세

| 타이밍 | 업데이트 대상 | 설명 |
|--------|-------------|------|
| `initialize_tree` 진입 | `base_layer._seen_tokens += prompt_delta_len` | Base KV cache에 prompt 반영 |
| `initialize_tree` 완료 | `cache.pending_draft_tokens`<br>`cache.retrieve_indices`<br>`cache.tree_mask`<br>`cache.tree_position_ids` | 첫 draft tree 메타데이터 저장 |
| `topk_generate` 내부 | `draft_layer._seen_tokens += input_ids_delta_len` | Draft KV cache에 토큰 반영 |
| `tree_decoding` 진입 시 accept_tokens 존재 | `base_layer._seen_tokens += accepted_prefix_length`<br>`cache.accept_tokens = None` | 이전 step의 accept를 base cache에 커밋 |
| `update_inference_inputs` | `cache.accept_tokens = accepted`<br>`cache.pending_draft_tokens = new_draft`<br>`cache.retrieve_indices = new_indices`<br>`cache.tree_mask = new_mask`<br>`cache.tree_position_ids = new_pos` | accept 토큰 저장 + 다음 tree 준비 |
| 종료 시 (`generate` 말미) | `cache.clear_tree_state()`<br>`cache.sync_draft_seq_length_to_base()` | tree 메타 정리, draft seq를 base에 동기화 |

### 구현 위치 요약

| 동작 | 파일 | 클래스/함수 |
|------|------|-----------|
| Base MXQ forward (chunked prefill/decode) | `eagle3_utils.py` | `MobilintEagle3BaseModelMixin.forward()` |
| Draft MXQ forward + tree 확장 | `eagle3_utils.py` | `MobilintEagle3DraftModelMixin.forward()`, `.topk_generate()` |
| FC 차원 변환 | `eagle3_utils.py` | `MobilintEagle3FCProjector.project()` |
| 초기 prefill + 첫 tree | `tree_decoding.py` | `initialize_tree()` |
| Tree verification | `tree_decoding.py` | `tree_decoding()` |
| Accept 길이 결정 | `tree_decoding.py` | `evaluate_posterior()` |
| 상태 갱신 + 다음 tree | `tree_decoding.py` | `update_inference_inputs()` |
| 메인 생성 루프 | `generation_utils.py` | `MobilintEagle3GenerationMixin._run_eagle3_decode_loop()` |
| Cache 관리 | `cache_utils.py` | `MobilintEagle3Cache` |
| Causal/tree mask 생성 | `eagle3_utils.py` | `make_causal_mask()`, `expand_mask()` |
| RoPE 테이블 사전 계산 | `eagle3_utils.py` | `CachedRotaryEmbedding`, `ScaledCachedRotaryEmbedding` |

### CODE REVIEW

**Comment by beomsu-mobilint** [comment]
@mobilint-review

**Comment by beomsu-mobilint** [comment]
@mobilint-review

**Comment by beomsu-mobilint** [comment]
@mobilint-review

**Comment by beomsu-mobilint** [comment]
@mobilint-review

**Comment by beomsu-mobilint** [comment]
@mobilint-review

**Comment by beomsu-mobilint** [comment]
@mobilint-review

**Comment by beomsu-mobilint** [comment]
@mobilint-review

**Comment by beomsu-mobilint** [comment]
@mobilint-review

**Comment by beomsu-mobilint** [comment]
@mobilint-review

**Comment by beomsu-mobilint** [comment]
@mobilint-review

**Comment by beomsu-mobilint** [comment]
@mobilint-review

**Comment by beomsu-mobilint** [comment]
@mobilint-review

**Comment by beomsu-mobilint** [comment]
@mobilint-review

**Comment by beomsu-mobilint** [comment]
@mobilint-review

**Comment by beomsu-mobilint** [comment]
@mobilint-review

**Comment by beomsu-mobilint** [comment]
@mobilint-review

**Comment by beomsu-mobilint** [comment]
@mobilint-review

**Comment by beomsu-mobilint** [comment]
@mobilint-review

**Comment by beomsu-mobilint** [comment]
@mobilint-review

**Comment by beomsu-mobilint** [comment]
@mobilint-review
